### PR TITLE
[FEATURE][2] Form validation

### PR DESCRIPTION
**Pull request by ziba-it**
# Form validation

## Description ⭐

I implemented _react-hook-form_ to handle and validate the sign up form.
This includes: custom validation rules, state management for loading and submission, and error handling to provide users with real-time feedback during the registration process.

## Screenshots 📷

Desktop screenshot

image

Mobile recording

https://github.com/user-attachments/assets/3077d5ac-e636-4d65-b9a4-ea42bc8b4a4f

## Links

- [Figma](https://www.figma.com/design/LWtEg9H7sdq1VrM7UqdyIq/JavaScript?node-id=213-22684&node-type=canvas&t=kVnauKGF0Rr9jeIB-0)

### CODE REVIEW

File: app/components/Button.tsx [hunk 1]
```diff
@@ -2,14 +2,15 @@ import { PropsWithChildren, ButtonHTMLAttributes } from "react";
 import { tv } from "tailwind-variants";
 
 type ButtonProps = {
-  variant: "primary";
+  variant: "primary" | "disabled";
```
Comment (hyanez-lightit) [inline app/components/Button.tsx]: disabled should not be a variant, rather a state of the button by the already-included `disabled` prop

File: app/components/Input.tsx [hunk 2]
```diff
@@ -2,28 +2,51 @@ import { ReactNode, InputHTMLAttributes } from "react";
...
   type,
   name,
+  register,
+  rules,
+  errors,
   value,
   ...rest
 }: InputProps) {
+  const inputState =
+    errors && errors[name] ? "error" : value ? "filled" : "empty";
```
Comment (hyanez-lightit) [inline app/components/Input.tsx]: [avoid nested ternaries](https://eslint.org/docs/latest/rules/no-nested-ternary)

File: app/components/SignUpForm.tsx [hunk 3]
```diff
@@ -3,43 +3,71 @@ import { useState } from "react";
...
-    }));
-  };
-
-  const handleSubmit = (): void => {
-    setLoading(true);
-    setTimeout(() => {
-      setCompleted(true);
-      setLoading(false);
-    }, 1000);
+  const signUpRules = {
```
Comment (hyanez-lightit) [inline app/components/SignUpForm.tsx]: This can be defined outside the component

File: app/components/SignUpForm.tsx [hunk 4]
```diff
@@ -72,38 +100,59 @@ export default function SignUpForm() {
...
+      <Button
+        variant={hasErrors ? "disabled" : "primary"}
+        disabled={hasErrors}
+        onClick={handleSubmit(() => {
+          setLoading(true);
+          setTimeout(() => {
+            setLoading(false);
+            setCompleted(true);
+          }, 1000);
+        })}
```
Comment (hyanez-lightit) [inline app/components/SignUpForm.tsx]: try to outsource to a function to avoid polluting the button onclick

File: app/components/SignUpForm.tsx [hunk 5]
```diff
@@ -3,43 +3,71 @@ import { useState } from "react";
...
-      setLoading(false);
-    }, 1000);
+  const signUpRules = {
+    email: {
+      required: { value: true, message: "Email is required" },
+      pattern: {
+        value: /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/,
+        message: "Invalid email address",
+      },
+    },
```
Comment (hyanez-lightit) [inline app/components/SignUpForm.tsx]: Could you take a look into [zod](https://zod.dev/), more specifically its [rhf resolver](https://github.com/react-hook-form/resolvers?tab=readme-ov-file#zod)

File: app/components/InputLabel.tsx [hunk 6]
```diff
@@ -1,9 +1,24 @@
...
+import { tv } from "tailwind-variants";
 
-export default function InputLabel({ children }: PropsWithChildren) {
-  return (
-    <label className="font-manrope text-primary-600 font-normal text-xs tracking-wide">
-      {children}
-    </label>
-  );
+type InputLabelProps = {
+  inputState: "error" | "filled" | "empty";
```
Comment (caroasti) [inline app/components/InputLabel.tsx]: suggestion: `inputState` -> `state` 
The component name (InputLabel) already implies that the state relates to an input so this change will make the code more concise while still being fully descriptive.

File: app/components/InputLabel.tsx [hunk 7]
```diff
@@ -1,9 +1,24 @@
...
+      empty: "",
+    },
+  },
+});
+
+export default function InputLabel({
+  inputState,
+  children,
+}: PropsWithChildren<InputLabelProps>) {
+  return <label className={label({ state: inputState })}>{children}</label>;
```
Comment (caroasti) [inline app/components/InputLabel.tsx]: with previous suggestion you could use the [shorthand property](https://ui.dev/shorthand-properties) {label({ state })}

File: app/components/SignUpForm.tsx (hunk 5 above)
Comment (caroasti) [reply app/components/SignUpForm.tsx]: +1 and you can use then the `type FormData = z.infer<typeof formDataSchema>;`
https://zod.dev/?id=type-inference

File: app/components/SignUpForm.tsx [hunk 8]
```diff
@@ -72,38 +100,59 @@ export default function SignUpForm() {
...
         <Input
           icon={<Lock />}
           label="Confirm Password"
           placeholder="Confirm your password"
           type="password"
           name="confirmPassword"
-          onChange={handleChange}
-          value={formData.confirmPassword}
+          register={register}
+          required={true}
```
Comment (caroasti) [inline app/components/SignUpForm.tsx]: I see you have repetead times `required={true}`
The required rule is already included in the rules object under the register function, and this handles the validation directly.

File: app/components/SignUpForm.tsx [hunk 9]
```diff
@@ -136,22 +143,15 @@ export default function SignUpForm() {
...
           register={register}
-          required={true}
           rules={signUpRules.confirmPassword}
           errors={errors}
           value={watch("confirmPassword")}
         />
       </form>
       <Button
-        variant={hasErrors ? "disabled" : "primary"}
+        variant={"primary"}
```
Comment (caroasti) [inline app/components/SignUpForm.tsx]: not necessary to have `{` ` } `

**Pull request merged**